### ISSUE TITLE
Add ability to replicate medic-users-meta database to PostgreSQL

### ISSUE THREAD

**Issue by michaelkohn** [medic/medic-couch2pg#54]
Now that we are storing telemetry data and feedback docs in `medic-users-meta`, we should make this data more accessible by either automatically replicating this database, or adding an environment variable that allows users to specify whether or not they want to replicate it (DEFAULT value if not provided could be NO). 

Instead of replicating this data to the `couchdb` table in PostgreSQL, we might consider replicating it to a separate table such as`couchdb_users_meta`.  While I don’t think it happens as much now, we’ve seen projects with 10s of thousands of feedback docs and it would be better to keep that data out of the main transaction table (`couchdb`).

**Comment by yrimal** [comment]
Trying to make requirement more clear here:

1. Replicate `medic-users-meta` to the `couchdb_users_meta` table on same Postgres database where `medic`  is replicated. 
2. Since we don't have enough projects in latest CHT core where `medic-users-meta` exists, it's unclear whether `read:message:` docs are merged from `medic-user-{username}-meta` to `medic-users-meta` or not.  Neither App Services, Nor R&L team currently need `read:message` at the moment. So, we don't need to sync read report to postgres that will save us time and space as well. 

CC: @kennsippell

**Comment by helizabetholsen** [comment]
Thanks for clarifying, @yrimal. This issue is definitely a priority for us to enable our further exploration of telemetry data, particularly as we look for cross-cutting impact metrics for both MM hosted and self-hosted partners.

**Comment by MaxDiz** [comment]
added to v3.11 based on roadmap planning decision

**Comment by yrimal** [comment]
I'd like to refrain from 

> 3. Ability to sync without using separate couch2pg process would be great.

And request that we make a separate process for syncing telemetry data. This thought comes considering that we might have partners coming in future who are self-hosted and are not willing to share real data or metrics but are open to share telemetry documents. In this situation, we'll be able to directly sync telemetry data from partner server to our rdbms monthly or in any specified period. 

Thoughts : @helizabetholsen , @yembrick  ?

**Comment by michaelkohn** [comment]
FWIW... that's actually how it works today. You can get telemetry and feedback docs, but you have to set up separate `couch2pg` process. The purpose of this ticket was to make it easier to get that data by allowing users to specify whether or not to include it in one `couch2pg` process so that you _don't_ have to set up a second one. 

RE: future partners only giving access to telemetry data, I'd imagine we would create a specific API (and user privelage) for that rather than using couch2pg.

**Comment by garethbowen** [comment]
There is potential PII in the feedback docs so I would not feel comfortable replicating them in raw form from a self-hosted partner. We should be able to get metadata like frequency and type of doc but we will be somewhat limited in what we can gather for self-hosted partners just like we are for data from the `medic` db.

**Comment by helizabetholsen** [comment]
Hi team! I'm checking in on this request. I'd love to know the timeline for this ticket and that we are still prioritizing adding these data to PostgreSQL before the end of the year. I know that @MaxDiz has a product initiative exploring these data and we are also looking into these data for impact monitoring on the R&L team, with @yrimal supporting.

Thanks!

**Comment by MaxDiz** [comment]
Hi @helizabetholsen this issue is in the current release and we are looking to have it out before year-end (contingent upon Angular dev work completion).

**Comment by helizabetholsen** [comment]
Awesome news! Thanks @MaxDiz for the update. I know that @ecsalomon @yrimal and @mouriceb are all interested in exploring these data in relation to our monitoring work -- and I'm sure the product team is as well in part of the telemetry initiative!

**Comment by kennsippell** [comment]
And app services is needing it to keep our apps healthy and to align our goals for 2021. Reiterating that this issue is our top ask to be released before end of year.

**Comment by garethbowen** [comment]
The workaround that @michaelkohn [mentioned earlier](https://github.com/medic/medic-couch2pg/issues/54#issuecomment-737225709) is to run two instances of couch2pg, one for each db.

For monitoring app health there is is also some of this data exposed through the monitoring API and while it's currently not enough to diagnose an issue it should be used to prompt app services to investigate the issue.

Neither of these are ideal workarounds and this issue is still a priority for us but I wanted to make sure they're documented here in case this ends up blocking someone.

**Comment by craig-landry** [comment]
Tentatively assigning to @njogz per interest, but this work is still currently behind other Angular migration tasks in priority.

**Comment by garethbowen** [comment]
This issue is chargeable to Project `214` in clicktime.

**Comment by michaelkohn** [comment]
I've drafted a few expectations in the table below, and have a few things to note:

1. The [ `.replications`](https://docs.communityhealthtoolkit.org/apps/reference/app-settings/replications/) setting is available between 3.5 and 3.9. I assume pre-3.5 the users DBs either didn't exist or there wasn't a way to replicate them through app_settings, but I don't know for sure so we should get clarity on that. Regardless, we should consider what needs to be tested for each variation (pre `3.5`, `3.5-3.9`, and `3.10+`) 
2. In the table below, I've specified the env variable be named `couch2pg_include_users_meta` but feel free to recommend something else
3. I've also specified TRUE/FALSE as possible values for this env variable but don't know if we prefer something else (yes/no, for example). 
4. Not including the new environment variable is treated as FALSE (ie, don't replicate) 
4. Perhaps we should call the new table `couchdb_users_meta`
5. Just like we do for the `medic` and `medic-sentinel` DBs, I'd imagine we store the `seq` and `source` in `couchdb_progress`


### Scenarios
---

CHT App Version | Action |[ `.replications`](https://docs.communityhealthtoolkit.org/apps/reference/app-settings/replications/) set up | `couch2pg_include_users_meta` | Expected result
-- | -- | -- | -- | --
3.10+ | Set up new medic-couch2pg | N/A | (not provided) | No changes
3.10+ | Set up new medic-couch2pg | N/A | TRUE | Creates new database table and replicates from `medic-users-meta`
3.10+ | Set up new medic-couch2pg | N/A | FALSE | No changes
3.10+ | Upgrade but do not change any env variables | N/A | N/A | No changes
3.10+ | Update existing medic-couch2pg to include new env variable | N/A | TRUE | Creates new database table and replicates from `medic-users-meta`
3.10+ | Update existing medic-couch2pg to include new env variable | N/A | FALSE | No changes
3.5 to 3.9.x | Set up new medic-couch2pg | Yes | (not provided) | No changes
3.5 to 3.9.x | Set up new medic-couch2pg | Yes | TRUE | Creates new database table and replicates from `medic-users-meta`
3.5 to 3.9.x | Set up new medic-couch2pg | Yes | FALSE | No changes
3.5 to 3.9.x | Do not change any env variables | Yes | N/A | No changes
3.5 to 3.9.x | Update existing medic-couch2pg to include new env variable | Yes | TRUE | Creates new database table and replicates from `medic-users-meta`
3.5 to 3.9.x | Update existing medic-couch2pg to include new env variable | Yes | FALSE | No changes
3.5 to 3.9.x | Set up new medic-couch2pg | No | (not provided) | No changes
3.5 to 3.9.x | Set up new medic-couch2pg | No | TRUE | No changes
3.5 to 3.9.x | Set up new medic-couch2pg | No | FALSE | No changes
3.5 to 3.9.x | Upgrade but do not change any env variables | No | N/A | No changes
3.5 to 3.9.x | Update existing medic-couch2pg env variables | No | TRUE | No changes
3.5 to 3.9.x | Update existing medic-couch2pg env variables | No | FALSE | No changes
Pre 3.5 | N/A (I think) |   |   |

**Comment by helizabetholsen** [comment]
Thank you! This is great, appreciate the table @michaelkohn

**Comment by yrimal** [comment]
> There is potential PII in the feedback docs so I would not feel comfortable replicating them in raw form from a self-hosted partner. We should be able to get metadata like frequency and type of doc but we will be somewhat limited in what we can gather for self-hosted partners just like we are for data from the `medic` db.

Queries and views produced would be dependent upon whether we want to identify and separate potential PII in the feedback doc as a part of this issue or we'll handle that issue separately. Based on @garethbowen 's comment, we should apply filter on raw data itself before pushing to couchdb, unlike current medic-couch2pg. 

I can provide some view structure thereafter.

**Comment by michaelkohn** [comment]
@yrimal The scope of this issue is to replicate data from `medic-users-meta` database to PostgreSQL using `medic-couch2pg` so that it is as easily accessible as data from the `medic` DB. 

For this issue, we are _**not**_  exposing this data anywhere else outside of the `medic-couch2pg` process. Since this data is only going to be exposed in PostgreSQL, we should **_not_** apply any filters on the `medic-users-meta` database, just as we don't apply any filters on the `medic` database (which is 100% PII).

Adding functionality to enable exposure and/or remote collection of telemetry data , if desired, should be raised as a separate issue (if it hasn't already).

**Comment by michaelkohn** [comment]
Also, just to be clear @yrimal re: your https://github.com/medic/medic-couch2pg/issues/54#issuecomment-688018183 as it relates to filtering, I don't expect there to be any `read:message` documents in `medic-users-meta`. The documentation as described [here](https://docs.communityhealthtoolkit.org/apps/guides/database/#medic-users-meta) mentions that sentinel only replicates `feedback` and `telemetry` docs to the `medic-users-meta` database. If we are seeing `read:message` docs in `medic-users-meta`, it is probably a bug.

**Comment by yrimal** [comment]
Hi @michaelkohn ,
I've added definition for two materialized views for this case. I'm fine with VIEW or MATERIALIZED view since this sync will happen only once a month. 

@kennsippell , Let me know if you'd like to add any other fields, which'll be of help for App-Service Team, on initial views that'll be created.

**Comment by garethbowen** [comment]
Ready for AT in...

couch2pg branch `54-specify-table-name-to-replicate-to`
medic-couch2pg branch `54-replicate-medic-users-meta-database`

Make sure the pg migration runs successfully, then the users-meta-data db in couchdb is replicated to the new `couchdb_users_meta` table in postgres. Work with Yuvraj to review the schema (tables and views) and make sure it's fit for purpose.

**Comment by yrimal** [comment]
Hi @njogz , 

Is this also the requirement for production enviornment or was committed by mistake ? I

https://github.com/medic/medic-couch2pg/blob/99b5ae1d39f6454952c532851254942fc49a75e5/package.json#L20

**Comment by njogz** [comment]
@yrimal I added that so that it would be easy to test the medic-couch2pg changes without being blocked by the couch2pg changes. That will be replaced by the appropriate couch2pg package before being merged.

**Comment by newtewt** [comment]
I see the data is being replicated to `couchdb_users_meta` now. 

@njogz  I have one call out that I'm not 100% is going to be an issue or not. As I don't know how the data is being consumed.  It seems the sql query is expecting the docs to always have an object as the `info` property. There are times where it is just a string. 
The query gets the message property from the info object. 

`  couchdb_users_meta.doc #>> '{info,message}'::text[] AS message`


When you log a bug report in the app. The query breaks because info is a string not an object.  I assume we should probably update webapp to form info always as an object instead of adding a bunch of sql logic.  The result of the `userview_feedback`  materialized view just returns null. The data is in the database. 

  "info": "a bug report?!",

**Comment by newtewt** [comment]
Another note, when there is telemetry sync that has a version set to a branch and not an integer. It will break the useview_telemetary query.

Err: ``` ERROR:  invalid input syntax for type integer: "6444-places-workflows-sms"
SQL state: 22P02```


   ```     CASE
            WHEN string_to_array(couchdb_users_meta.doc #>> '{metadata,versions,app}'::text[], '.'::text)::integer[] < string_to_array('3.8.0'::text, '.'::text)::integer[] THEN (couchdb_users_meta.doc #>> '{metadata,month}'::text[])::integer
            ELSE ((couchdb_users_meta.doc #>> '{metadata,month}'::text[])::integer) + 1
        END, '-1')::date AS period_start,```

**Comment by newtewt** [comment]
and I crashed `medic-couch2pg` in the docker container. 

```
[2021-03-22T17:59:18.393Z ERROR]:  { error: invalid input syntax for type integer: "6444-places-workflows-sms"
    at Connection.parseE (/app/node_modules/pg/lib/connection.js:545:11)
    at Connection.parseMessage (/app/node_modules/pg/lib/connection.js:370:19)
    at Socket.<anonymous> (/app/node_modules/pg/lib/connection.js:113:22)
    at Socket.emit (events.js:198:13)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:269:11)
    at Socket.Readable.push (_stream_readable.js:224:10)
    at TCP.onStreamRead [as onread] (internal/stream_base_commons.js:94:17)
  name: 'error',
  length: 256,
  severity: 'ERROR',
  code: '22P02',
  detail: undefined,
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where:
   'SQL statement "REFRESH MATERIALIZED VIEW CONCURRENTLY useview_telemetry"\nPL/pgSQL function refresh_matviews() line 15 at EXECUTE',
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'numutils.c',
  line: '259',
  routine: 'pg_strtoint32' }
[2021-03-22T17:59:18.394Z ERROR]:  An unrecoverable error occurred
[2021-03-22T17:59:18.394Z ERROR]:  TypeError: Assignment to constant variable.
    at run (/app/libs/medic/runner.js:55:20)
    at process._tickCallback (internal/process/next_tick.js:68:7)
[2021-03-22T17:59:18.394Z ERROR]:  exiting
```

**Comment by njogz** [comment]
@yrimal mind helping me fix the views?

**Comment by yrimal** [comment]
Hi @newtewt ,
I think the version number being a branch name like `6444-places-workflows-sms` in production is unlikely AFAIK the versioning system of medic cht-core version. This field is basically pulling `cht-core`'s version number and one variant I'm seeing is the version like `3.10.2-beta-1` and I can update the query to handle version number like this. 

If there is any chance of version number being in non-standard version format, we'll need to fallback to the query not fixing the issue [6604](https://github.com/medic/cht-core/issues/6604) and [5949](https://github.com/medic/cht-core/issues/5949). That version check is actually needed to handle problems caused by 6604 and 5949. If we switch to not handling that, we need to handle it in similar way after couch2pg is run and also let any partners using this feature know that there exists an issue like this and that needs to be handled. 

CC: @garethbowen / @helizabetholsen . 

There's bit of a  [discussion](https://medic.slack.com/archives/C024KTGRW/p1598071435009600) around this.

**Comment by dianabarsan** [comment]
> production is unlikely AFAIK

We have production instances on custom branches (like Muso).

**Comment by yrimal** [comment]
@newtewt , I've pushed the changes to fix crashing issue and handle other two issues separately on the branch. You should be good to continue testing.

**Comment by garethbowen** [comment]
The build failed based on too many logins to Docker - this is fixed in a separate PR so we can ignore this for now. Moving back to ready for AT.

**Comment by newtewt** [comment]
I'm seeing the fix for the branches and the crash no longer happening. However, the original issue I mentioned is still an issue. To me a change in the webapp to structure all feedback docs the same, but this means pre-whatever fix version it would show as null in some cases. 

Examples some are 
```
{
  "info": "this is a bug!",
 }
 ```
 and others are
 ```
 {
   "info": {
    "message": "Uncaught (in promise): {\"name\":\"Error\",\"message\":\"\"}",
    "stack": "Error: Uncaught (in promise): {\"name\":\"Error\",\"message\":\"\"}\n    at E (https://192-168-1-215.my.local-ip.co:444/polyfills.js?_sw-precache=76abd25b3b164a682f388e7a347960ef:1:19840)\n    at E (https://192-168-1-215.my.local-ip.co:444/polyfills.js?_sw-precache=76abd25b3b164a682f388e7a347960ef:1:19364)\n    at https://192-168-1-215.my.local-ip.co:444/polyfills.js?_sw-precache=76abd25b3b164a682f388e7a347960ef:1:20640\n    at e.value (https://192-168-1-215.my.local-ip.co:444/polyfills.js?_sw-precache=76abd25b3b164a682f388e7a347960ef:1:10733)\n    at Object.onInvokeTask (https://192-168-1-215.my.local-ip.co:444/main.js?_sw-precache=5d0d62839d9c420925ea94d070662378:1:901943)\n    at e.value (https://192-168-1-215.my.local-ip.co:444/polyfills.js?_sw-precache=76abd25b3b164a682f388e7a347960ef:1:10654)\n    at t.value (https://192-168-1-215.my.local-ip.co:444/polyfills.js?_sw-precache=76abd25b3b164a682f388e7a347960ef:1:4915)\n    at _ (https://192-168-1-215.my.local-ip.co:444/polyfills.js?_sw-precache=76abd25b3b164a682f388e7a347960ef:1:13024)"
  },
 }
 
 ```

>  It seems the sql query is expecting the docs to always have an object as the info property. There are times where it is just a string.
The query gets the message property from the info object.

> ` couchdb_users_meta.doc #>> '{info,message}'::text[] AS message`

> When you log a bug report in the app. The query breaks because info is a string not an object. I assume we should probably update webapp to form info always as an object instead of adding a bunch of sql logic. The result of the userview_feedback materialized view just returns null. The data is in the database.

> `"info": "a bug report?!",`

**Comment by michaelkohn** [comment]
Personally, i would recommend having a consistent data structure produced by cht-core. Something like:

```JSON
{
  "info": {
   "message": "An error occurred",
   "debug": "A stack trace that normally looks like a stack trace but is omitted for brevity"
 },
}
```

```JSON
{
  "info": {
   "message": "missing",
   "debug": "Possibly unhandled rejection: {\"status\":404,\"name\":\"not_found\",\"message\":\"missing\",\"error\":true}"
 },
}
```

```JSON
{
  "info": {
   "message": "this is a bug!",
   "debug": ""
 },
}
```

And the view we build can easily have the logic built into it to map data into that format if it is from an earlier version.

**Comment by dianabarsan** [comment]
> i would recommend having a consistent data structure produced by cht-core

even if we do this, the view should be compatible with older data, which clearly is not using the suggested format.

**Comment by yrimal** [comment]
@newtewt ,
I've updated the mat view creation logic, so that it understands that we've some inconsistency on writing feedback docs. If `{info,cause}` is NULL, it'll pull `{info}` as a cause; which in the instances I'm testing is working.

**Comment by dianabarsan** [comment]
To keep things transparent, we're waiting for feedback from @michaelkohn about a recent push to the PR branch. Related commit and message: https://github.com/medic/medic-couch2pg/commit/4e2e044cbe8d9f52fc19ff5ef9321697548cca56#commitcomment-49150811 (now orphaned because of rebase) https://github.com/medic/medic-couch2pg/pull/69/commits/8fbfefef6e57971f07bd5d6ee59198bdb764d74b

**Comment by yrimal** [comment]
@kennsippell , @derickl ,
You might have some feedback on what fields and information you want to be included on the materialized view that's created with this implementation. We can always create views that suffice our needs, but we're trying to create a [view](https://github.com/medic/medic-couch2pg/blob/63d46470fc74d9767ab76e1c5da9e808507ff864/libs/medic-users-meta/migrations/202102191153.do.54-create-couchdb-users-meta-table.sql#L3-L16) that's inclusive of basic needs.

**Comment by michaelkohn** [comment]
After chatting with @derickl, here is my recommendation for `feedback` docs.

```DROP MATERIALIZED VIEW IF EXISTS useview_feedback;

CREATE MATERIALIZED VIEW useview_feedback AS
(
	SELECT 
	    doc->>'_id' AS uuid,
	    doc#>>'{meta,source}' AS source,	    
	    doc#>>'{meta,app}' AS app,
	    doc#>>'{meta,version}' AS app_version,
	    doc#>>'{meta,user,name}' AS username,
	    doc#>>'{meta,url}' AS url,
	    doc->'info' AS info,
	    doc#>>'{info,cause}' AS cause,
	    doc#>>'{info,stack}' AS stack,
	    doc#>>'{info,message}' AS message,
	    doc->'log' AS log,
	    doc#>>'{meta,time}' AS time

	FROM
	    couchdb_users_meta

	WHERE
	    doc->>'type'='feedback'
);

CREATE UNIQUE INDEX idx_useview_feedback_uuid ON useview_feedback(uuid);
```

For `telemetry` docs.  Since there is no way to definitively know whether or not the telemetry doc has [the date bug](https://github.com/medic/medic-couch2pg/issues/54#issuecomment-804959439), I don't think this generic view should attempt to address it. 

Also, since the list of `metrics` are dynamic (there will be additional keys per `enketo:reports:<form>:<action>:<component>`), I've added an additional view that handles that by creating one row per metric.

```DROP VIEW IF EXISTS telemetryview_metrics;
DROP MATERIALIZED VIEW IF EXISTS useview_telemetry;

CREATE MATERIALIZED VIEW useview_telemetry AS
(
	SELECT 
		doc->>'_id' AS uuid,
		doc#>>'{metadata,year}' AS year,
		doc#>>'{metadata,month}' AS month,
		doc#>>'{metadata,versions,app}' AS app_version,		
		doc#>>'{metadata,user}' AS username,
		doc#>>'{metadata,versions,forms}' AS form_versions,
		doc#>'{metrics}' AS metrics,
		doc#>'{device}' AS device,
		doc#>'{dbInfo}' AS dbinfo

	FROM
	    couchdb_users_meta
	    
	WHERE
	    doc->>'type'='telemetry'
);

CREATE UNIQUE INDEX idx_useview_telemetry_uuid ON useview_telemetry(uuid);

CREATE OR REPLACE VIEW telemetryview_metrics AS
(
	WITH metrics_objects AS
	(
		SELECT
			uuid,
			year,
			month,
			user_name,
			app_version,
			row_to_json(jsonb_each(metrics)) AS metrics
		FROM
			useview_telemetry
	)

	SELECT
		uuid,
		year,
		month,
		user_name,
		app_version,
		metrics->>'key' AS metric,
		metrics#>>'{value,sum}' AS sum,
		metrics#>>'{value,min}' AS min,
		metrics#>>'{value,max}' AS max,
		metrics#>>'{value,count}' AS count,
		metrics#>>'{value,sumsqr}' AS sumsqr

	FROM
		metrics_objects
)	
;
 ```

**Comment by derickl** [comment]
We should figure out a way to extract the necessary data for `telemetryview_metrics` in one go without resorting the temparary materialisation to disk courtesy of the CTE @michaelkohn

**Comment by michaelkohn** [comment]
I'm not too concerned with that. This query will not be run very often and once we upgrade to PG 12, it won't automatically materialize this CTE (since it is used only once and doesn't include INSERT, UPDATE, or DELETE).

**Comment by michaelkohn** [comment]
I'm also fine omitting this view in efforts to not hold up 3.11 any longer.

**Comment by yrimal** [comment]
I am fine with that as well and we can safely exclude the version checking for telemetry docs if [6604](https://github.com/medic/cht-core/issues/6604), currently scheduled for 3.12 is out with update script for previous versions. 

For internal purpose, we can create view and start working on dashboards. And after the release of 6604, we can create generic views for both feedback docs and telemetry docs.

**Comment by derickl** [comment]
We can keep the view as is and figure out how to make it faster down the road

**Comment by dianabarsan** [comment]
I don't believe so. From my understanding, we're keeping the view as is.

**Comment by yrimal** [comment]
Let's move ahead with no changes made after last commit I made. The original purpose of the issue is already solved and we can talk separately and create views tha covers most of the scenario.

**Comment by newtewt** [comment]
ok cool, I'm seeing the varying cases of handling info whether it is a string or object.  

![image](https://user-images.githubusercontent.com/10531066/114448353-6fa42380-9ba1-11eb-87c5-69c235c52876.png)


and the telemetry being synced as well. 

![image](https://user-images.githubusercontent.com/10531066/114448488-9eba9500-9ba1-11eb-8972-4f0b0a3e53fb.png)

**Comment by dianabarsan** [comment]
Merged to `master`.
